DatadogSessionReplay/Sources/Recorder/RecordingTrigger.swift

```diff
@@ -0,0 +1,71 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+#if os(iOS)
+import DatadogInternal
+import UIKit
+
+/// Orchestrates the process of triggering next snapshot recording.
+/// It has a couple recording triggers which are done by using swizzling.
+internal protocol RecordingTriggering {
+    func startWatchingTriggers(_ callback: @escaping () -> Void)
+    func stopWatchingTriggers()
+}
+
+internal final class RecordingTrigger: RecordingTriggering, UIViewHandler, UIEventHandler, CALayerHandler {
+    private var uiViewSwizzler: UIViewSwizzler? = nil
+    private var uiApplicationSwizzler: UIApplicationSwizzler? = nil
+    private var caLayerSwizzler: CALayerSwizzler? = nil
+
+    private var triggerCallback: (() -> Void)?
+
+    init() throws {
+        uiViewSwizzler = try UIViewSwizzler(handler: self)
+        uiApplicationSwizzler = try UIApplicationSwizzler(handler: self)
+        caLayerSwizzler = try CALayerSwizzler(handler: self)
+    }
+
+    deinit {
+        uiViewSwizzler?.unswizzle()
+        uiApplicationSwizzler?.unswizzle()
+        caLayerSwizzler?.unswizzle()
+    }
+
+    func startWatchingTriggers(_ callback: @escaping () -> Void) {
+        triggerCallback = callback
+        uiViewSwizzler?.swizzle()
+        uiApplicationSwizzler?.swizzle()
+        caLayerSwizzler?.swizzle()
+    }
+
+    func stopWatchingTriggers() {
+        triggerCallback = nil
+        uiViewSwizzler?.unswizzle()
+        uiApplicationSwizzler?.unswizzle()
+        caLayerSwizzler?.unswizzle()
+    }
+
+    func notify_setNeedsDisplay(layer: CALayer) {
+        triggerCallback?()
+    }
+
+    func notify_draw(layer: CALayer, context: CGContext) {
+        triggerCallback?()
+    }
+
+    func notify_layoutSubviews(view: UIView) {
+        triggerCallback?()
+    }
+
+    func notify_viewDidAppear(viewController: UIViewController) {
+        triggerCallback?()
+    }
+
+    func notify_sendEvent(application: UIApplication, event: UIEvent) {
+        triggerCallback?()
+    }
```

Comment:
Could elaborate on why these 5 methods? The PR description only mention 2 and I don't see why we would need `notify_viewDidAppear` or `notify_setNeedsDisplay` vs. `notify_draw` vs. `notify_layoutSubviews` 🤔

Reply:
I cleaned it up. I was playing with many different permutations and this one seems the most stable.

Although I'll give it another shot with Maxime's suggestion below to trigger the notification before the implementation.